src/core/device.c

```diff
@@ -718,153 +741,176 @@ static bool device_is_ready(sd_device *dev) {
-                        continue;
+                if (devname && path_equal(d->path, devname))
+                        continue; /* This is the real device node. */
 
-                if (sd_device_get_syspath(assigned, &s) < 0)
-                        continue;
+                if (device_has_devlink(dev, d->path))
+                        continue; /* The devlink was already processed in the above loop. */
 
-                if (path_equal(s, syspath))
-                        continue;
+                if (strv_contains(aliases, d->path))
+                        continue; /* This is already processed in the above, and ready. */
 
-                r = strv_extend(&ready_devlinks, d->path);
-                if (r < 0)
-                        return -ENOMEM;
+                if (path_startswith(d->path, "/dev/"))
+                        /* This is a devlink unit. Check existence and update syspath. */
+                        (void) device_setup_devlink_unit_one(m, d->path, ready_units, not_ready_units);
+                else
+                        /* This is an alias unit of dropped or not ready device. */
+                        (void) set_ensure_put(not_ready_units, NULL, d);
         }
 
-        *ret_ready_devlinks = TAKE_PTR(ready_devlinks);
         return 0;
 }
 
-static void device_process_new(Manager *m, sd_device *dev, const char *sysfs) {
-        const char *dn, *alias;
+static int device_setup_units(Manager *m, sd_device *dev, Set **ready_units, Set **not_ready_units) {
+        const char *syspath, *devname = NULL;
         int r;
 
         assert(m);
         assert(dev);
-        assert(sysfs);
+        assert(ready_units);
+        assert(not_ready_units);
 
-        /* Add the main unit named after the sysfs path. If this one fails, don't bother with the rest, as
-         * this one shall be the main device unit the others just follow. (Compare with how
-         * device_following() is implemented, see below, which looks for the sysfs device.) */
-        if (device_setup_unit(m, dev, sysfs, /* main = */ true) < 0)
-                return;
+        r = sd_device_get_syspath(dev, &syspath);
+        if (r < 0)
+                return log_device_debug_errno(dev, r, "Couldn't get syspath from device, ignoring: %m");
 
-        /* Add an additional unit for the device node */
-        if (sd_device_get_devname(dev, &dn) >= 0)
-                (void) device_setup_unit(m, dev, dn, /* main = */ false);
+        /* First, process the main (that is, points to the syspath) and (real, not symlink) devnode units. */
+        if (device_for_action(dev, SD_DEVICE_REMOVE))
+                /* If the device is removed, the main and devnode units units will be removed by
+                 * device_update_found_by_sysfs() in device_dispatch_io(). Hence, it is not necessary to
+                 * store them to not_ready_units, and we have nothing to do here.
```

Comment:
`device_update_found_by_sysfs()` is not called in `device_enumerate()`, but I assume that this condition can only be triggered from `device_dispatch_io()` and not `device_enumerate()`?

Reply:
No removed devices can be enumerated. Hence, that's no problem.